### ISSUE TITLE
Modify interlocked nodes to take an indirection

### ISSUE THREAD

**Issue by CarolEidt** [dotnet/coreclr#14557]
The op1 of an interlocked instruction is an address. The instructions implicitly do an indirection. This makes sense because interlocked instructions are fundamentally operands on  memory locations. However, this complicates codegen. First, because op1 doesn't hold the type of the memory location (see issues #14539), the type must either be on the interlocked instruction, which is problematic if the node isn't actually producing a value, OR we must use the type of op2 (which is what is currently done for XARCH). Second, because `emitter::emitInsBinary()` expects to see an indirection, we create one to pass to that method (https://github.com/dotnet/coreclr/blob/master/src/jit/codegenxarch.cpp#L3703).
It would probably be better to simply redefine these nodes to take a value (i.e. an indirection of a memory location) instead of an address.

**Comment by CarolEidt** [comment]
As @sdmaclea pointed out, there may be an issue for 64 bit atomics if the addend is an indirect small type, when there should be a carry into the high 32 bits.

We should improve the test coverage especially for 64-bit atomics.

**Comment by mikedn** [comment]
> It would probably be better to simply redefine these nodes to take a value (i.e. an indirection of a memory location) instead of an address.

That's an interesting situation. Changing the first operand to be an indir seems easy but does this imply that these nodes should also produce a value that is consumed by a separate `GT_STOREIND`?

**Comment by CarolEidt** [comment]
> Changing the first operand to be an indir seems easy but does this imply that these nodes should also produce a value that is consumed by a separate GT_STOREIND?

Hmmm - I don't think that would be great, as it would appear to be separate ops. Since the JIT doesn't directly model dataflow of memory (aside from non-address-exposed locals), I guess I would probably not directly model the store (i.e. the side-effect of the atomic op would be to modify the location of memory that it loads).

**Comment by mikedn** [comment]
Right, VN already treats GT_CMPXCHG specially https://github.com/dotnet/coreclr/blob/e69300f4c06684b90f93ea9dfa1ce98c0b71626d/src/jit/valuenum.cpp#L7094-L7097

It may be nice to avoid such special cases but it's not clear if it has any real benefit. It certainly makes things more complicated, at least for as long as we have assignments in IR.

**Comment by mikedn** [comment]
> First, because op1 doesn't hold the type of the memory location

This issue seems to be specific to `GT_LOCKADD`. And then it turns out that `GT_LOCKADD` is kind of pointless - there are no actual `Interlocked` methods that return `void` and `GT_LOCKADD` nodes are only created by a rather weird optimization. 

"Weird" because it happens very early and it caters to ISA specifics. It should be done during lowering and then we my find out that `GT_LOCKADD` isn't really needed. LSRA/CodeGen could probably recognize an unused `GT_XADD` node and do the right thing. While I tend to favor lowering to IR that is similar to the actual assembly code in this case I'd say that doing this is overkill.

**Comment by mikedn** [comment]
It's interesting to note that currently interlocked nodes fail to contain address modes:
```
488D4204             lea      rax, bword ptr [rdx+4]
F0                   lock
830001               add      dword ptr [rax], 1
```
Probably a side-effect of their special nature...

**Comment by CarolEidt** [comment]
Closed, as #18267 addressed the issues in a different way.
